homeassistant/loader.py

```diff
@@ -1132,43 +1130,39 @@ def bind_hass(func: _CallableT) -> _CallableT:
 
 async def _async_component_dependencies(
     hass: HomeAssistant,
-    start_domain: str,
     integration: Integration,
-    loaded: set[str],
-    loading: set[str],
 ) -> set[str]:
-    """Recursive function to get component dependencies.
+    """Async function to get component dependencies."""
+    loading = set()
+    loaded = set()
 
-    Async friendly.
-    """
-    domain = integration.domain
-    loading.add(domain)
+    async def helper(integration: Integration) -> None:
+        """Recursive async helper to get component dependencies."""
+        domain = integration.domain
+        loading.add(domain)
 
-    for dependency_domain in integration.dependencies:
-        # Check not already loaded
-        if dependency_domain in loaded:
-            continue
+        for dependency_domain in integration.dependencies:
+            dep_integration = await async_get_integration(hass, dependency_domain)
 
-        # If we are already loading it, we have a circular dependency.
-        if dependency_domain in loading:
-            raise CircularDependency(domain, dependency_domain)
+            # If we are already loading it, we have a circular dependency.
+            # We have to check it here, because we need to cover all possible dependants.
+            if conflict := loading.intersection(dep_integration.after_dependencies):
+                raise CircularDependency(conflict.pop(), dependency_domain)
 
-        loaded.add(dependency_domain)
-
-        dep_integration = await async_get_integration(hass, dependency_domain)
+            # If we have already loaded it, no point doing it again.
+            if dependency_domain in loaded:
+                continue
 
-        if start_domain in dep_integration.after_dependencies:
-            raise CircularDependency(start_domain, dependency_domain)
+            # If we are already loading it, we have a circular dependency.
+            if dependency_domain in loading:
+                raise CircularDependency(dependency_domain, domain)
```

Comment:
Why are the arguments swapped?

Reply:
I'm not sure what exactly do you mean. Swapped compared to what?
The reason it is in this order is because `dependency_domain` is already in `loading`, therefore it was encountered earlier in the dependency graph than `domain`.

Reply:
I mean that the arguments to the exception are swapped:

```diff
-raise CircularDependency(domain, dependency_domain)
+raise CircularDependency(dependency_domain, domain)
```

The exception is intended to have a descriptive message, does it still make sense after swapping the arguments?
```py
class CircularDependency(LoaderError):
    """Raised when a circular dependency is found when resolving components."""

    def __init__(self, from_domain: str, to_domain: str) -> None:
        """Initialize circular dependency error."""
        super().__init__(f"Circular dependency detected: {from_domain} -> {to_domain}.")
```

Reply:
> The exception is intended to have a descriptive message, does it still make sense after swapping the arguments?

Yes, it does. It is a cycle after all, one integration depends on the other, which in turn depends on the first one.

In fact the current order is the order in which it makes more sense in my view. `dependency_domain` is encountered earlier in the dependency chain than `domain`, so it is in fact better to log it this way.

Unless I'm misunderstanding the error message, which is possible.

Reply:
I am not strongly attached to this order though, so if you wish to have it swapped, I can do that.